### ISSUE TITLE
Trying to use the package leads to a max-lisp-eval-depth error

### ISSUE THREAD

**Issue by safijari** [copilot-emacs/copilot.el#326]
I'm getting the following error (alongside CPU getting saturated for a bit) when trying to activate copilot with this package. Emacs 29.3 installed on NixOs.

```
advice-add: Lisp nesting exceeds ‘max-lisp-eval-depth’: 1601
Error running timer ‘copilot--post-command-debounce’: (excessive-lisp-nesting 1601)
```

My config is 
```
(use-package copilot
  :straight (:host github :repo "copilot-emacs/copilot.el" :files ("dist" "*.el"))
  :ensure t)

  (setq copilot-enable-delay 1.00)
  (define-key copilot-completion-map (kbd "<return>") 'copilot-accept-completion)
  (define-key copilot-completion-map (kbd "<backspace>") 'copilot-clear-overlay)
  (global-set-key (kbd "C-SPC") 'copilot-complete)
  (setq copilot-node-executable "/run/current-system/sw/bin/node")
```

Would appreciate guidance.

**Comment by corpix** [comment]
ok, this happens to me and I am on NixOS too.

TL;DR: NixOS specific issue, use `copilot-language-server-fhs` (should be available in unstable) https://github.com/NixOS/nixpkgs/pull/393260 or try https://nix.dev/permalink/stub-ld

here is trace from debugger

```
Debugger entered--Lisp error: (excessive-lisp-nesting 1601)
  (rassq (or mode major-mode) major-mode-remap-alist)
  (car (rassq (or mode major-mode) major-mode-remap-alist))
  (and (boundp 'major-mode-remap-alist) (car (rassq (or mode major-mode) major-mode-remap-alist)))
  (list (or mode major-mode) (and (boundp 'major-mode-remap-alist) (car (rassq (or mode major-mode) major-mode-remap-alist))))
  (append (if mode nil yas--extra-modes) (list (or mode major-mode) (and (boundp 'major-mode-remap-alist) (car (rassq (or mode major-mode) major-mode-remap-alist)))))
  (remq nil (append (if mode nil yas--extra-modes) (list (or mode major-mode) (and (boundp 'major-mode-remap-alist) (car (rassq (or mode major-mode) major-mode-remap-alist))))))
  (delete-dups (remq nil (append (if mode nil yas--extra-modes) (list (or mode major-mode) (and (boundp 'major-mode-remap-alist) (car (rassq (or mode major-mode) major-mode-remap-alist)))))))
  (let* ((modes (delete-dups (remq nil (append (if mode nil yas--extra-modes) (list (or mode major-mode) (and ... ...))))))) (yas--merge-ordered-lists (mapcar #'yas--all-parents modes)))
  yas--modes-to-activate()
  (let ((tail (yas--modes-to-activate))) (while tail (let ((mode (car tail))) (let ((funs (reverse (gethash mode yas--scheduled-jit-loads)))) (let ((tail funs)) (while tail (let (...) (yas--message 4 "Loading for `%s', just-in-time: %s!" mode fun) (funcall fun) (setq tail ...)))) (remhash mode yas--scheduled-jit-loads)) (setq tail (cdr tail)))))
  yas--load-pending-jits()
  (cond ((and yas-minor-mode (featurep 'yasnippet)) (if (memql 'yas--direct-keymaps emulation-mode-map-alists) (with-no-warnings emulation-mode-map-alists) (setq emulation-mode-map-alists (cons 'yas--direct-keymaps emulation-mode-map-alists))) (add-hook 'post-command-hook #'yas--post-command-handler nil t) (let ((tail (yas--modes-to-activate))) (while tail (let ((mode (car tail))) (let ((name ...)) (set-default name nil) (set (make-local-variable name) t)) (setq tail (cdr tail))))) (yas--load-pending-jits) (yas--auto-fill-wrapper) (add-hook 'auto-fill-mode-hook #'yas--auto-fill-wrapper)) (t (remove-hook 'post-command-hook #'yas--post-command-handler t) (remove-hook 'auto-fill-mode-hook #'yas--auto-fill-wrapper) (if (local-variable-p 'auto-fill-function) (progn (let* ((new (advice--remove-function ... ...))) (if (eq new (advice--buffer-local ...)) nil (advice--set-buffer-local 'auto-fill-function new))))) (setq emulation-mode-map-alists (remove 'yas--direct-keymaps emulation-mode-map-alists))))
  (let ((last-message (current-message))) (setq yas-minor-mode (cond ((eq arg 'toggle) (not yas-minor-mode)) ((and (numberp arg) (< arg 1)) nil) (t t))) (if (boundp 'local-minor-modes) (progn (setq local-minor-modes (delq 'yas-minor-mode local-minor-modes)) (if yas-minor-mode (progn (setq local-minor-modes (cons 'yas-minor-mode local-minor-modes)))))) (cond ((and yas-minor-mode (featurep 'yasnippet)) (if (memql 'yas--direct-keymaps emulation-mode-map-alists) (with-no-warnings emulation-mode-map-alists) (setq emulation-mode-map-alists (cons 'yas--direct-keymaps emulation-mode-map-alists))) (add-hook 'post-command-hook #'yas--post-command-handler nil t) (let ((tail (yas--modes-to-activate))) (while tail (let ((mode ...)) (let (...) (set-default name nil) (set ... t)) (setq tail (cdr tail))))) (yas--load-pending-jits) (yas--auto-fill-wrapper) (add-hook 'auto-fill-mode-hook #'yas--auto-fill-wrapper)) (t (remove-hook 'post-command-hook #'yas--post-command-handler t) (remove-hook 'auto-fill-mode-hook #'yas--auto-fill-wrapper) (if (local-variable-p 'auto-fill-function) (progn (let* ((new ...)) (if (eq new ...) nil (advice--set-buffer-local ... new))))) (setq emulation-mode-map-alists (remove 'yas--direct-keymaps emulation-mode-map-alists)))) (run-hooks 'yas-minor-mode-hook (if yas-minor-mode 'yas-minor-mode-on-hook 'yas-minor-mode-off-hook)) (if (called-interactively-p 'any) (progn nil (if (and (current-message) (not (equal last-message (current-message)))) nil (let ((local " in current buffer")) (message "%s %sabled%s" "yas minor mode" (if yas-minor-mode "en" "dis") local))))))
  yas-minor-mode(1)
  (if (or (not (or (listp yas-dont-activate-functions) (functionp yas-dont-activate-functions))) (run-hook-with-args-until-success 'yas-dont-activate-functions)) nil (yas-minor-mode 1))
  yas-minor-mode-on()
  (if yas-minor-mode (progn (yas-minor-mode -1) (yas-minor-mode-on)) (yas-minor-mode-on))
  (if (eq yas-minor-mode-major-mode major-mode) nil (if yas-minor-mode (progn (yas-minor-mode -1) (yas-minor-mode-on)) (yas-minor-mode-on)))
  (if yas-minor-mode-set-explicitly nil (if (eq yas-minor-mode-major-mode major-mode) nil (if yas-minor-mode (progn (yas-minor-mode -1) (yas-minor-mode-on)) (yas-minor-mode-on))))
  yas-global-mode-enable-in-buffer()
  run-hooks(after-change-major-mode-hook)
  run-mode-hooks()
  fundamental-mode()
  jsonrpc--forwarding-buffer("*copilot stderr*" "[stderr] " #<jsonrpc-process-connection jsonrpc-process-connection-30db7cc>)
  #f(compiled-function (conn slots) #<bytecode -0x17d0085d54906f23>)(#<jsonrpc-process-connection jsonrpc-process-connection-30db7cc> (:name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0)))
  apply(#f(compiled-function (conn slots) #<bytecode -0x17d0085d54906f23>) (#<jsonrpc-process-connection jsonrpc-process-connection-30db7cc> (:name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0))))
  #f(compiled-function (&rest args) #<bytecode -0x13eac5a4697f67eb>)(#<jsonrpc-process-connection jsonrpc-process-connection-30db7cc> (:name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0)))
  apply(#f(compiled-function (&rest args) #<bytecode -0x13eac5a4697f67eb>) #<jsonrpc-process-connection jsonrpc-process-connection-30db7cc> (:name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0)))
  initialize-instance(#<jsonrpc-process-connection jsonrpc-process-connection-30db7cc> (:name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0)))
  #f(compiled-function (class &rest slots) "Default constructor for CLASS `eieio-default-superclass'.\nSLOTS are the initialization slots used by `initialize-instance'.\nThis static method is called when an object is constructed.\nIt allocates the vector used to represent an EIEIO object, and then\ncalls `initialize-instance' on that object." #<bytecode -0x45a1c1768425257>)(jsonrpc-process-connection :name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0))
  apply(#f(compiled-function (class &rest slots) "Default constructor for CLASS `eieio-default-superclass'.\nSLOTS are the initialization slots used by `initialize-instance'.\nThis static method is called when an object is constructed.\nIt allocates the vector used to represent an EIEIO object, and then\ncalls `initialize-instance' on that object." #<bytecode -0x45a1c1768425257>) jsonrpc-process-connection (:name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0)))
  make-instance(jsonrpc-process-connection :name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0))
  apply(make-instance (jsonrpc-process-connection :name "copilot" :request-dispatcher copilot--handle-request :notification-dispatcher copilot--handle-notification :process #<process copilot server> :events-buffer-config (:size 0)))
  #f(compiled-function (&rest args2) #<bytecode -0x11ada60bcce63274>)(:events-buffer-config (:size 0))
  copilot--make-connection()
  copilot--start-server()
...
```

disabling `yas` doesn't make much difference, because:

> `*copilot stderr*`

```
Could not start dynamically linked executable: /run/current-system/sw/bin/copilot-language-server
NixOS cannot run dynamically linked executables intended for generic
linux environments out of the box. For more information, see:
https://nix.dev/permalink/stub-ld

Process copilot server stderr finished
```

looks like https://github.com/NixOS/nixpkgs/issues/391730

```
 ~  λ  ldd /run/current-system/sw/bin/copilot-language-server                                                                                                                                                                                  
        linux-vdso.so.1 (0x00007ffff7fc3000)
        libdl.so.2 => /nix/store/cmpyglinc9xl9pr4ymx8akl286ygl64x-glibc-2.40-66/lib/libdl.so.2 (0x00007ffff7fb8000)
        libstdc++.so.6 => not found
        libm.so.6 => /nix/store/cmpyglinc9xl9pr4ymx8akl286ygl64x-glibc-2.40-66/lib/libm.so.6 (0x00007ffff7ecf000)
        libgcc_s.so.1 => /nix/store/4vm6fj41qg4b4m99zlklvcxwfg6jvrja-xgcc-14-20241116-libgcc/lib/libgcc_s.so.1 (0x00007ffff7ea1000)
        libpthread.so.0 => /nix/store/cmpyglinc9xl9pr4ymx8akl286ygl64x-glibc-2.40-66/lib/libpthread.so.0 (0x00007ffff7e9c000)
        libc.so.6 => /nix/store/cmpyglinc9xl9pr4ymx8akl286ygl64x-glibc-2.40-66/lib/libc.so.6 (0x00007ffff7c94000)
        /lib64/ld-linux-x86-64.so.2 => /nix/store/cmpyglinc9xl9pr4ymx8akl286ygl64x-glibc-2.40-66/lib64/ld-linux-x86-64.so.2 (0x00007ffff7fc5000)
```

probably a most simple solution would be to have FHS version of the language server... package maintainer added `copilot-language-server-fhs` recentry https://github.com/NixOS/nixpkgs/pull/393260

**Comment by bbatsov** [comment]
This is a NixOS-specific issue caused by dynamic linking problems with the copilot-language-server binary. As identified in the comments, the solution is to use `copilot-language-server-fhs` from nixpkgs (available in unstable) or use the stub-ld approach documented at https://nix.dev/permalink/stub-ld.

This is not a copilot.el issue. Closing as resolved.
